expressvpn@Hukuta/files/expressvpn@Hukuta/applet.js

```diff
@@ -0,0 +1,105 @@
+const Lang = imports.lang;
+const Applet = imports.ui.applet;
+const GLib = imports.gi.GLib;
+const Settings = imports.ui.settings;
+const MainLoop = imports.mainloop;
+const UUID = "expressvpn@Hukuta";
+
+
+function MyApplet(metadata, orientation, instance_id) {
+    this._init(metadata, orientation, instance_id);
+}
+
+MyApplet.prototype = {
+    __proto__: Applet.IconApplet.prototype,
+
+    _init: function(metadata, orientation, instance_id) {
+        Applet.IconApplet.prototype._init.call(this, orientation);
+
+        this.server = "smart";
+        this.is_connected = false;
+        this.process = null;
+        this.icon_connected = metadata.path + '/icon.png';
+        this.icon_disconnected = metadata.path + '/icon-grey.png';
+
+        try {
+            this.settings = new Settings.AppletSettings(this, UUID, instance_id);
+            this.settings.bindProperty(Settings.BindingDirection.IN, "server", "server", this._onServerChange);       
+            this.set_applet_icon_path(this.icon_disconnected);
+            this._onServerChange();
+            this.actor.connect('button-release-event', Lang.bind(this, this._onButtonReleaseEvent));
+            this._getConnectionStatusSecondary();
+        }
+        catch (e) {
+            global.logError(e);
+        }
+        
+    },
+
+    _onServerChange: function() {        
+        this._getConnectionStatus()
+    },
+
+    _onButtonReleaseEvent: function(actor, event) {
+        if (this._applet_enabled && this.process == null) {
+            if (event.get_button() == 1) {
+                if (!this._draggable.inhibit) {
+                    return false;
+                } else {
+                    let command = "expressvpn " + (this.is_connected ? "disconnect" : "connect " + this.server);                    
+                    this.process = MainLoop.timeout_add(2500, Lang.bind(this, this._onTimerAction));
+                    this.set_applet_icon_name("network-vpn");
+                    this.set_applet_tooltip("Please wait...");
+                    GLib.spawn_command_line_async(command, this._getConnectionStatus);
+                }
+            }
+        }
+        return true;
+    },
+
+    _onTimerAction: function() {
+        if (this._getConnectionStatus()) {
+            MainLoop.source_remove(this.process);
+            this.process = null;
+        } else {
+            this.process = MainLoop.timeout_add(2500, Lang.bind(this, this._onTimerAction));
+        }
+    },
+
+    _getConnectionStatus: function() {
+        let statusStr = ("" + GLib.spawn_command_line_sync("expressvpn status")[1]).split("\n")[0].replace(
```

Comment:
Hi, sorry for the delay. It might be a good idea to let the user know they need to install the expressvpn client if they don't have it.

```
Cjs-Message: 04:58:37.689: JS LOG: [LookingGlass/info] Installing settings for expressvpn@Hukuta
Cjs-Message: 04:58:37.690: JS LOG: [LookingGlass/info] Settings successfully installed for expressvpn@Hukuta
Cjs-Message: 04:58:37.698: JS LOG: [LookingGlass/error] Failed to execute child process “expressvpn” (No such file or directory)
Cjs-Message: 04:58:37.698: JS LOG: [LookingGlass/trace] 
<----------------
anonymous/MyApplet.prototype._getConnectionStatus@/home/jason/.local/share/cinnamon/applets/expressvpn@Hukuta/applet.js:70:31
anonymous/MyApplet.prototype._onServerChange@/home/jason/.local/share/cinnamon/applets/expressvpn@Hukuta/applet.js:40:9
anonymous/MyApplet.prototype._init@/home/jason/.local/share/cinnamon/applets/expressvpn@Hukuta/applet.js:29:13
MyApplet@/home/jason/.local/share/cinnamon/applets/expressvpn@Hukuta/applet.js:10:5
main@/home/jason/.local/share/cinnamon/applets/expressvpn@Hukuta/applet.js:103:20
createApplet@/usr/share/cinnamon/js/ui/appletManager.js:581:18
addAppletToPanels@/usr/share/cinnamon/js/ui/appletManager.js:356:22
finishExtensionLoad@/usr/share/cinnamon/js/ui/appletManager.js:95:14
Extension.prototype._init/<@/usr/share/cinnamon/js/ui/extension.js:277:17
---------------->
```

Reply:
What is a good way to do this? Is there any generally accepted design? Example?

Reply:
I'd say the easiest way is to handle the Exception is in the catch statement at Line 33.

In there you can check if the exception matches what @jaszhix has shown above, and display "You need to install expressvpn to use this applet" or sth on your applet.

Assuming the Error traversed up to that try/catch (I did not try out your code). There are probably several ways to handle it, this is just one of them.

Example:
``` Javascript
catch (e) {
     if (e.message == "Error you want to match it to") {
          // User facing message somewhere
          this.set_applet_tooltip("You need to install expressvpn to use this applet");
     }
     global.logError(e);
 }

PS.: (You probably need to handle the same error when the user clicks on your applet as well)

Reply:
Added exception handling and tooltip